test/e2e/framework/node/init/init.go

```diff
@@ -31,13 +32,9 @@ func init() {
 	framework.NewFrameworkExtensions = append(framework.NewFrameworkExtensions,
 		func(f *framework.Framework) {
 			ginkgo.AfterEach(func(ctx context.Context) {
-				if f.ClientSet == nil {
-					// Test didn't reach f.BeforeEach, most
-					// likely because the test got
-					// skipped. Nothing to check...
-					return
-				}
-				e2enode.AllNodesReady(ctx, f.ClientSet, 7*time.Minute)
+				ginkgo.DeferCleanup(func() {
```

Comment:
I'm reading the docs https://onsi.github.io/ginkgo/#spec-cleanup-aftereach-and-defercleanup and I don't understand from there how is this supposed to work , to use DeferCleanup inside AfterEach sounds at least very hard to reason about

Reply:
I agree, it's a bit hard to understand, but essentially this code schedules node readiness check  a bit further in time, after all other deferred cleanup callbacks. From my point of view it makes perfect sense and gives test cases an ability to restore node state in their deferred cleanup callback if they change it for the test purposes.

Reply:
One thing in the above mentioned doc that worth pointing out here:
```
Under the hood DeferCleanup is generating a dynamic AfterEach node and adding it to the running spec. 
```
So, this code generates dynamic `AfterEach`, which will be called later, allowing `DeferCleanup`s created by the test case to do their work.

Reply:
@bart0sh: this is not what we discussed in our chat last week. There I suggested 
```
ginkgo.BeforeEach(func() {
    gingko.DeferCleanup(func() {
        e2enode.AllNodesReady(ctx, f.ClientSet, 7*time.Minute)
    })
})
```
as replacement for the current `AfterEach`.

With my proposal, the `DeferCleanup` gets called before any code provided by a test, and because of the first-in-last-out order for `DeferCleanup`, the test is able to do whatever is necessary to make nodes ready again in its own `DeferCleanup`.

With your combination of `BeforeEach` + `DeferCleanup`, the `DeferCleanup` gets callback gets register *after* the test's callback and then gets invoked *before* the test's cleanup - not what we want to achieve.